### DISCUSSION TITLE
PNPM warning about configuration in a `Nx` repo

### DISCUSSION THREAD

**Discussion by zetashift** [aws/aws-pdk#490] [Q&A] [answered]
Since a `projen upgrade` I've been getting the following error:

```
runtime/typescript      |  WARN  The field "pnpm" was found in /api/runtime/typescript/package.json. This will not take effect. You should configure "pnpm" at the root of the workspace instead.
````

This warning popped up in all of my subprojects, including the one generated by `type-safe-api`. Anybody know how I could fix this warning? 

Using
- `@aws-prototyping-sdk/nx-monorepo 0.19.33`
- `projen 0.71.120`
- `pnpm 8.6.5`

**Comment by agdimech** [comment]
Can you show me your package.json file and also are you actually using pnpm or yarn?

**Reply by zetashift** [reply]
```json
﻿
{
  "name": "mic-api-typescript-runtime",
  "description": "Generated typescript packages for our API, from Smithy models",
  "scripts": {
    "build": "npx projen build",
    "compile": "npx projen compile",
    "default": "npx projen default",
    "generate": "npx projen generate",
    "package": "npx projen package",
    "post-compile": "npx projen post-compile",
    "post-upgrade": "npx projen post-upgrade",
    "pre-compile": "npx projen pre-compile",
    "test": "npx projen test",
    "upgrade": "npx projen upgrade",
    "watch": "npx projen watch"
  },
  "devDependencies": {
    "@types/node": "^16",
    "npm-check-updates": "^16",
    "projen": "^0.71.20",
    "typescript": "^4.9.5"
  },
  "dependencies": {
    "@types/aws-lambda": "^8.10.114"
  },
  "pnpm": {},
  "main": "lib/index.js",
  "license": "Apache-2.0",
  "version": "0.0.0",
  "types": "lib/index.d.ts",
  "//": "~~ Generated by projen. To modify, edit .projenrc.js and run \"npx projen\"."
}
```

I am actually using pnpm, I don't have any `yarn` files and my projenrc contains:

```javascript
﻿const monorepo = new NxMonorepoProject({
  defaultReleaseBranch: "main",
  devDeps: [
    "@aws-prototyping-sdk/nx-monorepo",
    "@aws-prototyping-sdk/type-safe-api",
  ],
  name: "micx",
  packageManager: javascript.NodePackageManager.PNPM,

  projenrcTs: true,
  // description: undefined,  /* The description is just a string that helps people understand the purpose of the package. */
  // packageName: undefined,  /* The "name" in package.json. */
  gitignore: [".direnv", ".devenv", ".vscode", ".envrc", ".env"],
});

// Other subprojects
  packageManager: javascript.NodePackageManager.PNPM,

```

**Comment by agdimech** [comment, marked as answer]
This seems to be due to how projen renders resolutions: https://github.com/projen/projen/blob/bdec85e18b6ec1043ef64816638741e4abeb2b6b/src/javascript/node-package.ts#L1357C37-L1357C43

In the current implementation all pnpm projects will have at a minimum a `pnpm` field. If resolutions are present then they will be rendered. This could be fixed by changing Projen's implementation to only render a `pnpm` field if overrides exist, however I don't think the effort is worth it just to remove a warning although completely your call :)

Another option I just thought of is you could actually just remove the `pnpm` fields by modifying the package.json of each subproject i.e:

```
monorepoProject.subprojects
  .filter((p) => p instanceof NodeProject && !!p.package.manifest.pnpm)
  .forEach((p) => delete p.package.manifest.pnpm);
```

**Reply by zetashift** [reply]
Ah I see! I don't mind the warning knowing it's just about the resolution rendering (and the workaround for removing the pnpm field is enough).

Since a recent upgrade pnpm has been giving me the following warning:
```
 WARN  Issues with peer dependencies found
.
├─┬ @aws-prototyping-sdk/nx-monorepo 0.19.62
│ └── ✕ unmet peer projen@^0.71.92: found 0.72.8
└─┬ @aws-prototyping-sdk/type-safe-api 0.19.62
  └── ✕ unmet peer projen@^0.71.92: found 0.72.8
  ```
  
Also looks like a non-issue to me?

**Reply by agdimech** [reply]
It’s because projen recently bumped their minor version to 72, but these deps still have peer dependencies set to 71 whilst only a 72 projen is detected.

mid you downgrade your installed version of projen to 71, this will resolve.

**Comment by agdimech** [comment]
Looks like the empty pnpm object issue has been resolved in Projen: https://github.com/projen/projen/pull/3072
